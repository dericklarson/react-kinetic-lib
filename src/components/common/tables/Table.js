import React, { Component } from 'react';
import t from 'prop-types';
import {
  compose,
  lifecycle,
  withHandlers,
  withProps,
  withState,
} from 'recompose';
import { List, Map } from 'immutable';
import isarray from 'isarray';
import { I18n } from '../../core/i18n/I18n';
import {
  DefaultTableLayout,
  DefaultHeader,
  DefaultHeaderRow,
  DefaultHeaderCell,
  DefaultTableBody,
  DefaultTableBodyRow,
  DefaultEmptyBodyRow,
  DefaultTableBodyCell,
  DefaultTableFooter,
  DefaultTableFooterRow,
  DefaultTableFooterCell,
} from './defaults';
import { connect, dispatch } from '../../../store';
import {
  configureTable,
  mountTable,
  unmountTable,
  isClientSide,
} from './Table.redux';
import generateKey from '../../../helpers/generateKey';

const KeyWrapper = ({ children }) => children;

const TableComponent = props => {
  if (props.configured) {
    const { children, components } = props;
    const table = buildTable(props);
    const filter = buildFilterControl(props);
    const pagination = buildPaginationControl(props);

    return children({
      table,
      filter,
      pagination,
    });
  }
  return null;
};

const buildFilterControl = ({ components, filterProps }) => {
  const FilterControl = components.FilterControl;
  return FilterControl ? <FilterControl {...filterProps} /> : null;
};

const hasPrevPage = (data, pageTokens, pageOffset) =>
  isClientSide(data) ? pageOffset > 0 : pageTokens.size !== 0;

const hasNextPage = (data, pageOffset, pageSize, currentPageToken) =>
  isClientSide(data)
    ? data
      ? pageOffset + pageSize < data.length
      : false
    : !!currentPageToken;

const buildPaginationControl = props => {
  const {
    tableKey,
    data,
    pageTokens,
    pageSize,
    pageOffset,
    nextPageToken,
    currentPageToken,
    components,
    paginationProps,
  } = props;
  const PaginationControl = components.PaginationControl;
  const prevPage = hasPrevPage(data, pageTokens, pageOffset)
    ? onPrevPage(tableKey)
    : null;

  const nextPage = hasNextPage(data, pageOffset, pageSize, currentPageToken)
    ? onNextPage(tableKey)
    : null;

  return PaginationControl ? (
    <PaginationControl prevPage={prevPage} nextPage={nextPage} />
  ) : null;
};

export const buildTable = props => {
  const TableLayout = props.components.TableLayout
    ? props.components.TableLayout
    : DefaultTableLayout;
  const header = buildTableHeader(props);
  const body = buildTableBody(props);
  const footer = buildTableFooter(props);

  return <TableLayout header={header} body={body} footer={footer} />;
};

export const buildTableHeader = props => {
  const Header = props.components.Header || DefaultHeader;
  const headerRow = buildTableHeaderRow(props);
  return props.omitHeader ? null : (
    <Header sorting={props.sorting} headerRow={headerRow} rows={props.rows} />
  );
};

export const buildTableHeaderRow = props => {
  const { components, rows, columns } = props;
  const HeaderRow = components.HeaderRow || DefaultHeaderRow;
  const columnHeaders = columns.map(buildTableHeaderCell(props));

  return <HeaderRow columnHeaders={columnHeaders} rows={rows.toJS()} />;
};

export const buildTableHeaderCell = props => (column, index) => {
  const {
    tableKey,
    components,
    sorting,
    sortColumn,
    sortDirection,
    rows,
  } = props;
  const { title, sortable = true } = column;
  const CustomHeaderCell = column.components
    ? column.components.HeaderCell
    : null;
  const HeaderCell =
    CustomHeaderCell || components.HeaderCell || DefaultHeaderCell;

  return (
    <KeyWrapper key={`column-${index}`}>
      <HeaderCell
        onSortColumn={onSortColumn(tableKey, column)}
        title={title}
        column={column}
        sortColumn={sortColumn}
        sortDirection={sortDirection}
        rows={rows.toJS}
      />
    </KeyWrapper>
  );
};

export const buildTableBody = props => {
  const Body = props.components.Body || DefaultTableBody;
  const tableRows = buildTableBodyRows(props);

  return <Body tableRows={tableRows} rows={props.rows.toJS()} />;
};

export const buildTableBodyRows = props => {
  const { components, rows, columns, emptyMessage } = props;
  const BodyRow = components.BodyRow || DefaultTableBodyRow;
  const EmptyBodyRow = components.EmptyBodyRow || DefaultEmptyBodyRow;

  const tableRows =
    rows.size > 0 ? (
      rows.map((row, index) => {
        const cells = buildTableBodyCells(props, row, index);
        return (
          <KeyWrapper key={`row-${index}`}>
            <BodyRow
              cells={cells}
              columns={columns}
              row={row}
              index={index}
              rows={rows.toJS()}
            />
          </KeyWrapper>
        );
      })
    ) : (
      <EmptyBodyRow
        columns={columns}
        rows={rows.toJS()}
        emptyMessage={emptyMessage}
      />
    );

  return tableRows;
};

export const buildTableBodyCells = (props, row, rowIndex) => {
  const { components, rows, columns } = props;
  return columns.map((column, index) => {
    const CustomBodyCell = column.components
      ? column.components.BodyCell
      : null;
    const BodyCell =
      CustomBodyCell || components.BodyCell || DefaultTableBodyCell;
    const value = row[column.value];

    return (
      <KeyWrapper key={`column-${index}`}>
        <BodyCell
          row={row}
          index={index}
          value={value}
          rows={rows.toJS()}
          column={column}
        />
      </KeyWrapper>
    );
  });
};

export const buildTableFooter = props => {
  const { components, rows, includeFooter } = props;
  const footerRow = buildTableFooterRow(props);
  const Footer = components.Footer || DefaultTableFooter;

  return includeFooter ? (
    <Footer rows={rows.toJS()} footerRow={footerRow} />
  ) : null;
};

export const buildTableFooterRow = props => {
  const { components, rows, columns } = props;
  const cells = buildTableFooterCells(props);
  const FooterRow = components.FooterRow || DefaultTableFooterRow;

  return <FooterRow cells={cells} />;
};

export const buildTableFooterCells = props => {
  const { components, columns, rows } = props;
  return columns.map((column, index) => {
    const CustomFooterCell = column.components
      ? column.components.FooterCell
      : null;
    const FooterCell =
      CustomFooterCell || components.FooterCell || DefaultTableFooterCell;

    return (
      <KeyWrapper key={`column-${index}`}>
        <FooterCell column={column} rows={rows.toJS()} />
      </KeyWrapper>
    );
  });
};

const onNextPage = tableKey => () => dispatch('NEXT_PAGE', { tableKey });
const onPrevPage = tableKey => () => dispatch('PREV_PAGE', { tableKey });
const onSortColumn = (tableKey, column) => () =>
  dispatch('SORT_COLUMN', { tableKey, column });

const mapStateToProps = state => (state, props) =>
  state.getIn(['tables', props.tableKey], Map()).toObject();

/**
 * @component
 */
const TableImpl = compose(
  connect(mapStateToProps),
  lifecycle({
    componentDidMount() {
      if (this.props.mounted && !this.props.configured) {
        configureTable(this.props);
      }
    },

    componentDidUpdate() {
      if (this.props.mounted && !this.props.configured) {
        configureTable(this.props);
      }
    },
  }),
)(TableComponent);

class Table extends Component {
  constructor(props) {
    super(props);
    const { components, tableKey, ...config } = props;
    this.components = components;
    this.config = config;
    this.auto = !tableKey;
    this.tableKey = tableKey || generateKey();
  }

  componentDidMount() {
    if (this.auto) {
      mountTable(this.tableKey);
    }
  }

  componentWillUnmount() {
    if (this.auto) {
      unmountTable(this.tableKey);
    }
  }

  render() {
    return (
      <TableImpl {...this.props} tableKey={this.tableKey} auto={this.auto}>
        {this.props.children}
      </TableImpl>
    );
  }
}

Table.propTypes = {
  /** Either an array of data (for client-side tables) or an object providing datasource configuration. */
  data: t.oneOfType([t.array, t.object]).isRequired,
  /**
   * Calculated row data from the datasource mechanism.
   * @ignore
   */
  rows: t.instanceOf(List),
  /** An array of the columns that will be rendered on the table. */
  columns: t.arrayOf(
    t.shape({
      /** The title that will be rendered in the header. */
      title: t.string,
      /** The value key that will be used to map the column to the data object. */
      value: t.string,
      /** Flag that determines if the column can be used as a filter. */
      filterable: t.bool,
      /** Flag that determines if the column is sortable.*/
      sortable: t.bool,
      /** Allows overriding the `HeaderCell`, `BodyCell`, and `FooterCell` for a given column. */
      components: t.shape({
        HeaderCell: t.func,
        BodyCell: t.func,
        FooterCell: t.func,
      }),
    }),
  ).isRequired,
  components: t.shape({
    /** Override the default table layout, analogous to `<table>`. */
    TableLayout: t.func,

    /** Override the table header, analogous to `<thead>`. */
    Header: t.func,
    /** Override the table header row, analogous to the `<tr>` in `<thead>`. */
    HeaderRow: t.func,
    /** Override the table header cells, analogous to `<th>`. */
    HeaderCell: t.func,

    /** Override the table body, analogous to `<tbody>`. */
    Body: t.func,
    /** Override the table rows, analogous to `<tr>`. */
    BodyRow: t.func,
    /** Override the table body row when there is no data. */
    EmptyBodyRow: t.func,
    /** Override the table cells, analogous to `<td>`. */
    BodyCell: t.func,

    /** Override the table footer, analogous to `<tfoot>`. */
    Footer: t.func,
    /** Override the table footer row, analogous to the `<tr>` in `<tfoot>`. */
    FooterRow: t.func,
    /** Override the table footer cells, analogous to `<td>`. */
    FooterCell: t.func,
  }),
  /** Override the text message in the empty body row when there is no data in the table. */
  emptyMessage: t.string,
  /** Flag to enable/disable pagination support. */
  pagination: t.bool,
  /** Set the page size when paginating data. */
  pageSize: t.number,
  /** Flag to enable/disable filtering support. */
  filtering: t.bool,
  /** Flag to enable/disable sorting support. */
  sorting: t.bool,
  /** The child of this component should be a function which renders the table layout. */
  children: t.func.isRequired,

  /** A flag used to suppress the rendering of the table header. */
  omitHeader: t.bool,
  /** A flag used to enable to rendering of the table footer. */
  includeFooter: t.bool,
};

const defaultProps = {
  components: {},
  rows: List([]),
  columns: [],
  emptyMessage: 'No data found.',

  filtering: true,
  sorting: true,
  pagination: true,
  pageSize: 25,
  omitHeader: false,
  includeFooter: false,
};

Table.defaultProps = defaultProps;

Table.configure = configureTable;
Table.mount = mountTable;
Table.unmount = unmountTable;

/**
 * @component
 */
export default Table;
