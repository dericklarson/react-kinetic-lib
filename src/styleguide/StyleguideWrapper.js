import React, { Fragment } from 'react';
import { KineticLib } from 'react-kinetic-lib';
// import 'bootstrap/scss/bootstrap.scss';
// import 'font-awesome/css/font-awesome.css';

// Mock the bundle object.
window.bundle = window.bundle || {};
window.bundle.apiLocation = '/fake-api';

export default ({ children }) => <Fragment><KineticLib>{children}</KineticLib></Fragment>