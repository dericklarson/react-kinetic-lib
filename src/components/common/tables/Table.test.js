import React from 'react';
import { KineticLib } from '@kineticdata/react';
import { mount, render } from 'enzyme';
import { List } from 'immutable';
import { users } from '../../../styleguide/fixtures';
import Table, {
  buildTable,
  buildTableHeader,
  buildTableHeaderRow,
  buildTableHeaderCell,
  buildTableBody,
  buildTableBodyRows,
  buildTableBodyCells,
  buildTableFooter,
  buildTableFooterRow,
  buildTableFooterCells,
} from './Table';

describe('<Table />', () => {
  describe('build methods', () => {
    let props;
    let data = [];
    let columns = [];

    beforeEach(() => {
      data = users(2);
      columns = [{ value: 'username', title: 'Username' }];
      props = {
        components: {},
        data,
        columns,
        rows: List(data),
      };
    });

    describe('#buildTable', () => {
      test('it renders normally', () => {
        const wrapper = render(<KineticLib>{buildTable(props)}</KineticLib>);

        expect(wrapper.is('table')).toBeTruthy();
        expect(wrapper.is('table.custom-table')).toBeFalsy();
      });

      test('it renders a custom table', () => {
        const TableLayout = () => <table className="custom-table" />;

        props.components.TableLayout = TableLayout;
        const wrapper = render(<KineticLib>{buildTable(props)}</KineticLib>);

        expect(wrapper.is('table.custom-table')).toBeTruthy();
      });
    });

    describe('#buildTableHeader', () => {
      test('it renders normally', () => {
        const wrapper = render(
          <KineticLib>{buildTableHeader(props)}</KineticLib>,
        );

        expect(wrapper.is('thead')).toBeTruthy();
        expect(wrapper.is('thead.custom-thead')).toBeFalsy();
      });

      test('it renders a custom thead', () => {
        const Header = () => <thead className="custom-thead" />;

        props.components.Header = Header;
        const wrapper = render(
          <KineticLib>{buildTableHeader(props)}</KineticLib>,
        );

        expect(wrapper.is('thead.custom-thead')).toBeTruthy();
      });
    });

    describe('#buildTableHeaderRow', () => {
      test('it renders normally', () => {
        const wrapper = render(
          <KineticLib>{buildTableHeaderRow(props)}</KineticLib>,
        );

        expect(wrapper.is('tr')).toBeTruthy();
        expect(wrapper.is('tr.custom-tr')).toBeFalsy();
      });

      test('it renders a custom thead', () => {
        const HeaderRow = () => <tr className="custom-tr" />;

        props.components.HeaderRow = HeaderRow;
        const wrapper = render(
          <KineticLib>{buildTableHeaderRow(props)}</KineticLib>,
        );

        expect(wrapper.is('tr.custom-tr')).toBeTruthy();
      });
    });

    describe('#buildTableHeaderCell', () => {
      test('it renders normally', () => {
        const column = columns[0];
        const wrapper = render(
          <KineticLib>{buildTableHeaderCell(props)(column, 0)}</KineticLib>,
        );

        expect(wrapper.is('th')).toBeTruthy();
        expect(wrapper.is('td.custom-td')).toBeFalsy();
      });

      test('it renders a custom th', () => {
        const column = columns[0];
        const HeaderCell = () => <th className="custom-th" />;
        props.components.HeaderCell = HeaderCell;

        const wrapper = render(
          <KineticLib>{buildTableHeaderCell(props)(column, 0)}</KineticLib>,
        );

        expect(wrapper.is('th.custom-th')).toBeTruthy();
      });

      test('it renders a custom th for a specific column', () => {
        const column = columns[0];
        const HeaderCell = () => <th className="custom-cell-th" />;
        column.components = { HeaderCell };

        const wrapper = render(
          <KineticLib>{buildTableHeaderCell(props)(column, 0)}</KineticLib>,
        );

        expect(wrapper.is('th.custom-cell-th')).toBeTruthy();
      });
    });

    describe('#buildTableBody', () => {
      test('it renders normally', () => {
        const wrapper = render(
          <KineticLib>{buildTableBody(props)}</KineticLib>,
        );

        expect(wrapper.is('tbody')).toBeTruthy();
        expect(wrapper.is('tbody.custom-tbody')).toBeFalsy();
      });

      test('it renders a custom tbody', () => {
        const Header = () => <tbody className="custom-tbody" />;

        props.components.Header = Header;
        const wrapper = render(
          <KineticLib>{buildTableHeader(props)}</KineticLib>,
        );

        expect(wrapper.is('tbody.custom-tbody')).toBeTruthy();
      });
    });

    describe('#buildTableBodyRows', () => {
      test('it renders rows normally', () => {
        const wrapper = render(
          <KineticLib>
            <table>
              <tbody>{buildTableBodyRows(props)}</tbody>
            </table>
          </KineticLib>,
        );

        expect(wrapper.find('tr')).toHaveLength(props.rows.size);
        expect(
          wrapper
            .find('tr')
            .first()
            .hasClass('custom-tr'),
        ).toBeFalsy();
      });

      test('it renders custom rows', () => {
        const BodyRow = () => <tr className="custom-tr" />;
        props.components.BodyRow = BodyRow;

        const wrapper = render(
          <KineticLib>
            <table>
              <tbody>{buildTableBodyRows(props)}</tbody>
            </table>
          </KineticLib>,
        );

        expect(wrapper.find('tr')).toHaveLength(props.rows.size);
        expect(
          wrapper
            .find('tr')
            .first()
            .hasClass('custom-tr'),
        ).toBeTruthy();
      });

      test('it renders default empty row', () => {
        props.rows = List([]);

        const wrapper = render(
          <KineticLib>
            <table>
              <tbody>{buildTableBodyRows(props)}</tbody>
            </table>
          </KineticLib>,
        );

        expect(wrapper.find('tr')).toHaveLength(1);
        expect(
          wrapper
            .find('tr')
            .first()
            .hasClass('custom-empty-tr'),
        ).toBeFalsy();
      });

      test('it renders custom empty row', () => {
        const EmptyBodyRow = () => <tr className="custom-empty-tr" />;
        props.rows = List([]);
        props.components.EmptyBodyRow = EmptyBodyRow;

        const wrapper = render(
          <KineticLib>
            <table>
              <tbody>{buildTableBodyRows(props)}</tbody>
            </table>
          </KineticLib>,
        );

        expect(wrapper.find('tr')).toHaveLength(1);
        expect(wrapper.find('tr').hasClass('custom-empty-tr')).toBeTruthy();
      });
    });

    describe('#buildTableBodyCells', () => {
      test('it renders cells normally', () => {
        const wrapper = render(
          <KineticLib>
            <table>
              <tbody>
                <tr>{buildTableBodyCells(props, props.rows.first(), 0)}</tr>
              </tbody>
            </table>
          </KineticLib>,
        );

        expect(wrapper.find('tr td')).toHaveLength(1);
        expect(
          wrapper
            .find('tr td')
            .first()
            .hasClass('custom-td'),
        ).toBeFalsy();
      });

      test('it renders custom cells', () => {
        const BodyCell = () => <td className="custom-td" />;
        // props.columns = [...props.columns, { value: 'displayName', title: 'DisplayName'}]
        props.components.BodyCell = BodyCell;

        const wrapper = render(
          <KineticLib>
            <table>
              <tbody>
                <tr>{buildTableBodyCells(props, props.rows.first(), 0)}</tr>
              </tbody>
            </table>
          </KineticLib>,
        );

        expect(wrapper.find('tr td')).toHaveLength(props.columns.length);
        expect(
          wrapper
            .find('tr td')
            .first()
            .hasClass('custom-td'),
        ).toBeTruthy();
      });

      test('it renders custom column cells', () => {
        const BodyCell = () => <td className="custom-td" />;
        props.columns = [
          ...props.columns,
          {
            value: 'displayName',
            title: 'DisplayName',
            components: { BodyCell },
          },
        ];

        const wrapper = render(
          <KineticLib>
            <table>
              <tbody>
                <tr>{buildTableBodyCells(props, props.rows.first(), 0)}</tr>
              </tbody>
            </table>
          </KineticLib>,
        );

        expect(wrapper.find('tr td')).toHaveLength(props.columns.length);
        expect(
          wrapper
            .find('tr td')
            .first()
            .hasClass('custom-td'),
        ).toBeFalsy();
        expect(
          wrapper
            .find('tr td')
            .last()
            .hasClass('custom-td'),
        ).toBeTruthy();
      });
    });

    describe('#buildTableFooter', () => {
      test('it does not render normally', () => {
        const wrapper = render(
          <KineticLib>{buildTableFooter(props)}</KineticLib>,
        );

        expect(wrapper.is('tfoot')).toBeFalsy();
      });

      test('it renders with includeFooter', () => {
        props.includeFooter = true;
        const wrapper = render(
          <KineticLib>{buildTableFooter(props)}</KineticLib>,
        );

        expect(wrapper.is('tfoot')).toBeTruthy();
        expect(wrapper.is('tfoot.custom-tfoot')).toBeFalsy();
      });

      test('it renders a custom tfoot', () => {
        const Footer = () => <tfoot className="custom-tfoot" />;
        props.includeFooter = true;
        props.components.Footer = Footer;

        const wrapper = render(
          <KineticLib>{buildTableFooter(props)}</KineticLib>,
        );

        expect(wrapper.is('tfoot.custom-tfoot')).toBeTruthy();
      });
    });

    describe('#buildTableFooterRow', () => {
      test('it renders normally', () => {
        const wrapper = render(
          <KineticLib>{buildTableFooterRow(props)}</KineticLib>,
        );

        expect(wrapper.is('tr')).toBeTruthy();
        expect(wrapper.is('tr.custom-tr')).toBeFalsy();
      });

      test('it renders a custom tr', () => {
        const FooterRow = () => <tr className="custom-tr" />;

        props.components.FooterRow = FooterRow;
        const wrapper = render(
          <KineticLib>{buildTableFooterRow(props)}</KineticLib>,
        );

        expect(wrapper.is('tr.custom-tr')).toBeTruthy();
      });
    });

    describe('#buildTableFooterCells', () => {
      test('it renders normally', () => {
        const wrapper = render(
          <KineticLib>{buildTableFooterCells(props)}</KineticLib>,
        );

        expect(wrapper.first().is('td')).toBeTruthy();
        expect(wrapper.first().is('td.custom-td')).toBeFalsy();
      });

      test('it renders a custom td', () => {
        const FooterCell = () => <td className="custom-td" />;
        props.components.FooterCell = FooterCell;

        const wrapper = render(
          <KineticLib>{buildTableFooterCells(props)}</KineticLib>,
        );

        expect(wrapper.is('td.custom-td')).toBeTruthy();
      });

      test('it renders a custom td for a specific column', () => {
        const FooterCell = () => <td className="custom-td" />;
        props.columns = [
          ...props.columns,
          {
            value: 'displayName',
            title: 'DisplayName',
            components: { FooterCell },
          },
        ];

        const wrapper = render(
          <KineticLib>
            <tfoot>
              <tr>{buildTableFooterCells(props)}</tr>
            </tfoot>
          </KineticLib>,
        );

        expect(wrapper.find('td')).toHaveLength(props.columns.length);
        expect(
          wrapper
            .find('td')
            .first()
            .hasClass('custom-td'),
        ).toBeFalsy();
        expect(
          wrapper
            .find('td')
            .last()
            .hasClass('custom-td'),
        ).toBeTruthy();
      });
    });
  });
});
