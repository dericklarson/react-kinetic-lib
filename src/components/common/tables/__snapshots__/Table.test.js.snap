// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> when Header is overridden it renders the same each time 1`] = `
<KineticLib>
  <Provider
    context={
      Object {
        "$$typeof": Symbol(react.context),
        "Consumer": Object {
          "$$typeof": Symbol(react.context),
          "_calculateChangedBits": null,
          "_context": [Circular],
        },
        "Provider": Object {
          "$$typeof": Symbol(react.provider),
          "_context": [Circular],
        },
        "_calculateChangedBits": null,
        "_currentRenderer": Object {},
        "_currentRenderer2": null,
        "_currentValue": undefined,
        "_currentValue2": undefined,
        "_threadCount": 0,
      }
    }
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(Symbol.observable): [Function],
      }
    }
  >
    <I18nProvider>
      <withProps(Connect(withHandlers(withHandlers(withHandlers(withHandlers(lifecycle(TableComponent)))))))
        auto={true}
        columns={
          Array [
            Object {
              "title": "Username",
              "value": "username",
            },
          ]
        }
        components={Object {}}
        data={
          Array [
            Object {
              "username": "user-1",
            },
            Object {
              "username": "user-2",
            },
          ]
        }
        emptyMessage="There are no data rows."
        filtering={true}
        pageSize={10}
        pagination={true}
        rows={Immutable.List []}
        sorting={true}
        tableKey="abc"
      >
        <Connect(withHandlers(withHandlers(withHandlers(withHandlers(lifecycle(TableComponent))))))
          auto={false}
          columns={
            Array [
              Object {
                "title": "Username",
                "value": "username",
              },
            ]
          }
          components={Object {}}
          data={
            Array [
              Object {
                "username": "user-1",
              },
              Object {
                "username": "user-2",
              },
            ]
          }
          emptyMessage="There are no data rows."
          filtering={true}
          pageSize={10}
          pagination={true}
          rows={Immutable.List []}
          sorting={true}
          tableKey="abc"
        >
          <withHandlers(withHandlers(withHandlers(withHandlers(lifecycle(TableComponent)))))
            auto={false}
            columns={
              Array [
                Object {
                  "title": "Username",
                  "value": "username",
                },
              ]
            }
            components={Object {}}
            data={
              Array [
                Object {
                  "username": "user-1",
                },
                Object {
                  "username": "user-2",
                },
              ]
            }
            dispatch={[Function]}
            emptyMessage="There are no data rows."
            filtering={true}
            pageSize={10}
            pagination={true}
            rows={Immutable.List []}
            sorting={true}
            tableKey="abc"
          >
            <withHandlers(withHandlers(withHandlers(lifecycle(TableComponent))))
              auto={false}
              buildTableBodyCells={[Function]}
              buildTableFooterCells={[Function]}
              buildTableHeaderCell={[Function]}
              columns={
                Array [
                  Object {
                    "title": "Username",
                    "value": "username",
                  },
                ]
              }
              components={Object {}}
              data={
                Array [
                  Object {
                    "username": "user-1",
                  },
                  Object {
                    "username": "user-2",
                  },
                ]
              }
              dispatch={[Function]}
              emptyMessage="There are no data rows."
              filtering={true}
              pageSize={10}
              pagination={true}
              rows={Immutable.List []}
              sorting={true}
              tableKey="abc"
            >
              <withHandlers(withHandlers(lifecycle(TableComponent)))
                auto={false}
                buildTableBodyCells={[Function]}
                buildTableBodyRows={[Function]}
                buildTableFooterCells={[Function]}
                buildTableFooterRow={[Function]}
                buildTableHeaderCell={[Function]}
                buildTableHeaderRow={[Function]}
                columns={
                  Array [
                    Object {
                      "title": "Username",
                      "value": "username",
                    },
                  ]
                }
                components={Object {}}
                data={
                  Array [
                    Object {
                      "username": "user-1",
                    },
                    Object {
                      "username": "user-2",
                    },
                  ]
                }
                dispatch={[Function]}
                emptyMessage="There are no data rows."
                filtering={true}
                pageSize={10}
                pagination={true}
                rows={Immutable.List []}
                sorting={true}
                tableKey="abc"
              >
                <withHandlers(lifecycle(TableComponent))
                  auto={false}
                  buildTableBody={[Function]}
                  buildTableBodyCells={[Function]}
                  buildTableBodyRows={[Function]}
                  buildTableFooter={[Function]}
                  buildTableFooterCells={[Function]}
                  buildTableFooterRow={[Function]}
                  buildTableHeader={[Function]}
                  buildTableHeaderCell={[Function]}
                  buildTableHeaderRow={[Function]}
                  columns={
                    Array [
                      Object {
                        "title": "Username",
                        "value": "username",
                      },
                    ]
                  }
                  components={Object {}}
                  data={
                    Array [
                      Object {
                        "username": "user-1",
                      },
                      Object {
                        "username": "user-2",
                      },
                    ]
                  }
                  dispatch={[Function]}
                  emptyMessage="There are no data rows."
                  filtering={true}
                  pageSize={10}
                  pagination={true}
                  rows={Immutable.List []}
                  sorting={true}
                  tableKey="abc"
                >
                  <lifecycle(TableComponent)
                    auto={false}
                    buildFilterControl={[Function]}
                    buildTable={[Function]}
                    buildTableBody={[Function]}
                    buildTableBodyCells={[Function]}
                    buildTableBodyRows={[Function]}
                    buildTableFooter={[Function]}
                    buildTableFooterCells={[Function]}
                    buildTableFooterRow={[Function]}
                    buildTableHeader={[Function]}
                    buildTableHeaderCell={[Function]}
                    buildTableHeaderRow={[Function]}
                    columns={
                      Array [
                        Object {
                          "title": "Username",
                          "value": "username",
                        },
                      ]
                    }
                    components={Object {}}
                    data={
                      Array [
                        Object {
                          "username": "user-1",
                        },
                        Object {
                          "username": "user-2",
                        },
                      ]
                    }
                    dispatch={[Function]}
                    emptyMessage="There are no data rows."
                    filtering={true}
                    pageSize={10}
                    pagination={true}
                    rows={Immutable.List []}
                    sorting={true}
                    tableKey="abc"
                  >
                    <TableComponent
                      auto={false}
                      buildFilterControl={[Function]}
                      buildTable={[Function]}
                      buildTableBody={[Function]}
                      buildTableBodyCells={[Function]}
                      buildTableBodyRows={[Function]}
                      buildTableFooter={[Function]}
                      buildTableFooterCells={[Function]}
                      buildTableFooterRow={[Function]}
                      buildTableHeader={[Function]}
                      buildTableHeaderCell={[Function]}
                      buildTableHeaderRow={[Function]}
                      columns={
                        Array [
                          Object {
                            "title": "Username",
                            "value": "username",
                          },
                        ]
                      }
                      components={Object {}}
                      data={
                        Array [
                          Object {
                            "username": "user-1",
                          },
                          Object {
                            "username": "user-2",
                          },
                        ]
                      }
                      dispatch={[Function]}
                      emptyMessage="There are no data rows."
                      filtering={true}
                      pageSize={10}
                      pagination={true}
                      rows={Immutable.List []}
                      sorting={true}
                      tableKey="abc"
                    >
                      <div>
                        <DefaultTableLayout
                          body={
                            <DefaultTableBody
                              rows={Array []}
                              tableRows={
                                <DefaultEmptyBodyRow
                                  columns={
                                    Array [
                                      Object {
                                        "title": "Username",
                                        "value": "username",
                                      },
                                    ]
                                  }
                                  emptyMessage="There are no data rows."
                                  rows={Array []}
                                />
                              }
                            />
                          }
                          footer={null}
                          header={
                            <DefaultHeader
                              headerRow={
                                <DefaultHeaderRow
                                  columnHeaders={
                                    Array [
                                      <KeyWrapper>
                                        <DefaultHeaderCell
                                          column={
                                            Object {
                                              "title": "Username",
                                              "value": "username",
                                            }
                                          }
                                          onSortColumn={[Function]}
                                          rows={[Function]}
                                          title="Username"
                                        />
                                      </KeyWrapper>,
                                    ]
                                  }
                                  rows={Array []}
                                />
                              }
                              rows={Immutable.List []}
                              sorting={true}
                            />
                          }
                        >
                          <table>
                            <DefaultHeader
                              headerRow={
                                <DefaultHeaderRow
                                  columnHeaders={
                                    Array [
                                      <KeyWrapper>
                                        <DefaultHeaderCell
                                          column={
                                            Object {
                                              "title": "Username",
                                              "value": "username",
                                            }
                                          }
                                          onSortColumn={[Function]}
                                          rows={[Function]}
                                          title="Username"
                                        />
                                      </KeyWrapper>,
                                    ]
                                  }
                                  rows={Array []}
                                />
                              }
                              rows={Immutable.List []}
                              sorting={true}
                            >
                              <thead
                                className="sortable"
                              >
                                <DefaultHeaderRow
                                  columnHeaders={
                                    Array [
                                      <KeyWrapper>
                                        <DefaultHeaderCell
                                          column={
                                            Object {
                                              "title": "Username",
                                              "value": "username",
                                            }
                                          }
                                          onSortColumn={[Function]}
                                          rows={[Function]}
                                          title="Username"
                                        />
                                      </KeyWrapper>,
                                    ]
                                  }
                                  rows={Array []}
                                >
                                  <tr>
                                    <KeyWrapper
                                      key="column-0"
                                    >
                                      <DefaultHeaderCell
                                        column={
                                          Object {
                                            "title": "Username",
                                            "value": "username",
                                          }
                                        }
                                        onSortColumn={[Function]}
                                        rows={[Function]}
                                        title="Username"
                                      >
                                        <th
                                          scope="col"
                                        >
                                          <a
                                            onClick={[Function]}
                                          >
                                            <I18n>
                                              <I18nTranslate
                                                context="shared"
                                                loadTranslations={[Function]}
                                                locale="en"
                                                translations={Immutable.Map {}}
                                              />
                                            </I18n>
                                          </a>
                                        </th>
                                      </DefaultHeaderCell>
                                    </KeyWrapper>
                                  </tr>
                                </DefaultHeaderRow>
                              </thead>
                            </DefaultHeader>
                            <DefaultTableBody
                              rows={Array []}
                              tableRows={
                                <DefaultEmptyBodyRow
                                  columns={
                                    Array [
                                      Object {
                                        "title": "Username",
                                        "value": "username",
                                      },
                                    ]
                                  }
                                  emptyMessage="There are no data rows."
                                  rows={Array []}
                                />
                              }
                            >
                              <tbody>
                                <DefaultEmptyBodyRow
                                  columns={
                                    Array [
                                      Object {
                                        "title": "Username",
                                        "value": "username",
                                      },
                                    ]
                                  }
                                  emptyMessage="There are no data rows."
                                  rows={Array []}
                                >
                                  <tr>
                                    <td
                                      className="text-center"
                                    >
                                      <em>
                                        <I18n>
                                          <I18nTranslate
                                            context="shared"
                                            loadTranslations={[Function]}
                                            locale="en"
                                            translations={Immutable.Map {}}
                                          />
                                        </I18n>
                                      </em>
                                    </td>
                                  </tr>
                                </DefaultEmptyBodyRow>
                              </tbody>
                            </DefaultTableBody>
                          </table>
                        </DefaultTableLayout>
                      </div>
                    </TableComponent>
                  </lifecycle(TableComponent)>
                </withHandlers(lifecycle(TableComponent))>
              </withHandlers(withHandlers(lifecycle(TableComponent)))>
            </withHandlers(withHandlers(withHandlers(lifecycle(TableComponent))))>
          </withHandlers(withHandlers(withHandlers(withHandlers(lifecycle(TableComponent)))))>
        </Connect(withHandlers(withHandlers(withHandlers(withHandlers(lifecycle(TableComponent))))))>
      </withProps(Connect(withHandlers(withHandlers(withHandlers(withHandlers(lifecycle(TableComponent)))))))>
    </I18nProvider>
  </Provider>
</KineticLib>
`;
